fs: Add configurable matching criteria for --track-renames

quexten opened the pull request:
This commit adds the --track-using-modtime flag which makes
rclone match renames using the file modification time instead
of file hashes. This enables support for tracking renames in
encrypted remotes. For two files to match, their modification
times must be within 1 millisecond of each other to account
for remote modification time inaccuracies.

Fixes #3696
Fixes #2721

#### What is the purpose of this change?
This commit adds the `--track-using-modtime` flag which makes
rclone match renames using file modification time. It makes
`--track-renames` work with encrypted destinations.

#### Was the change discussed in an issue or in the forum before?
It was discussed [here](https://forum.rclone.org/t/can-not-use-track-renames-when-syncing-to-and-from-crypt-remotes/11063).

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

quexten commented:
As a note, for the test I just re-used the `--track-renames` test and modified it so it uses the new flag. 
I could use some help with the failing  build on the "modules_race" pipeline, as it builds fine on my repository.

ncw commented:
Thanks for taking a look at this and apologies for the delay looking at it.

I quite like the idea of making this a bit more general - see the comment here: https://forum.rclone.org/t/can-not-use-track-renames-when-syncing-to-and-from-crypt-remotes/11063/8

If we implemented a format string effectively for the rename hash then it would probably end up simplifying the code.

I note that the modtime comparison should be done according to the precision specified by `fs.GetModifyWindow` with the source and the dest Fs as parameters. Storing the modtime rendered as a unix time, divided by the precision (perhaps again divided by 2 as the comparison normally done is +/- precision) would be sufficient for matching I think.

quexten commented:
@ncw 
Thanks for the response!
I changed the parameter to now be a comma-separated list of requested strategies and moved the functions for generating the hash-string and the modtime string into 1 function which now just checks which strategies are requested and generates a comma separated string. So for the strategies hash,modtime it would be `size,hash,modtime`. I also changed the modtime comparison to use the precision as you suggested.
Let me know what you think.

ncw commented on fs/sync/sync.go:
```diff
@@ -557,20 +560,44 @@ func (s *syncCopyMove) srcParentDirCheck(entry fs.DirEntry) {
-		return ""
+
+	if isUsingRenameStrategy("modtime", renameStrategy) {
+		modTime := obj.ModTime(s.ctx).Unix() / precision.Nanoseconds() / 2
```
If `precision.Nanoseconds()` is small (eg it is 1 on local unix drives) then it would be nice to include the nanoseconds too in the comparison.

So the calculation really wants to be something like this (with rounding)

```go
modTime := obj.ModTime(s.ctx)
divisor := precision.Nanoseconds() / 2
rounding := divisor / 2
if rounding > 0 {
    rounding--
}
timeHash := (modTime.Unix() * time.Nanosecond + modTime.Nanosecond() + rounding ) / divisor
```

Note that `modTime.Unix() * time.Nanosecond` potentially overflows an int64. It will do so when modTime.Unix > 2**63/1E9 = the year 2262 so that is probably OK... If it isn't we'll have to break out the large integer maths library.

Note also we should probably be rejecting "modtime" if the `Precision == fs.ModTimeNotSupported`

quexten replied:
Okay, I have pushed another commit which resolves the issues you have pointed out. For the last suggestion, rejecting the modtime, I have handled it as it was previously handled when hashes are not available in either remote. So if one remote doesn't support modtime, (the window is ModTimeNotSupported) then `s.trackRenames` will be set to false.

ncw commented on fs/config/configflags/configflags.go:
```diff
@@ -61,6 +61,7 @@ func AddFlags(flagSet *pflag.FlagSet) {
 	flags.BoolVarP(flagSet, &deleteAfter, "delete-after", "", false, "When synchronizing, delete files on destination after transferring (default)")
 	flags.Int64VarP(flagSet, &fs.Config.MaxDelete, "max-delete", "", -1, "When synchronizing, limit the number of deletes")
 	flags.BoolVarP(flagSet, &fs.Config.TrackRenames, "track-renames", "", fs.Config.TrackRenames, "When synchronizing, track file renames and do a server side move if possible")
+	flags.StringVarP(flagSet, &fs.Config.TrackRenamesStrategy, "track-renames-strategy", "", fs.Config.TrackRenamesStrategy, "When tracking renames, use the modtime for comparison instead of the file hash")
```
The help for this is outdated

quexten replied:
Okay, updated the help in a new commit.

ncw commented:
That looks great - thank you very much :-)

I'll merge it now.

Mushin commented:
Will the --track-renames-strategy flag obey --modify-window ?

I'm trying to rename many files that will then update on an encrypted Gdrive. The mod-times at some point got changed, and are off by about a second, so I've been using --size-only. 

So I thought if I use the --track-renames with --modify-window, I could still rename the files server side?

ncw commented:
> Will the --track-renames-strategy flag obey --modify-window ?

Yes it will.

However I discovered a bug in the implementation after the integration tests failed: 93f5125f51866791b1a3f0d0e316123967db1453

So you might want to try [the latest beta](https://beta.rclone.org) in 15-30 mins.

Mushin commented:
> > Will the --track-renames-strategy flag obey --modify-window ?
> 
> Yes it will.
> 
> However I discovered a bug in the implementation after the integration tests failed: [93f5125](https://github.com/rclone/rclone/commit/93f5125f51866791b1a3f0d0e316123967db1453)
> 
> So you might want to try [the latest beta](https://beta.rclone.org) in 15-30 mins.

Great - thanks.
